src/algebra/homology/homology.lean

```diff
@@ -83,53 +103,53 @@ begin
   rw [has_zero_morphisms.comp_zero, image.fac_assoc, d_squared],
 end
 
--- TODO (a good project!):
--- At this level of generality, it's just not true that a chain map
--- induces maps on boundaries
---
--- Let's add these later, with appropriate (but hopefully fairly minimal)
--- assumptions: perhaps that the category is regular?
--- I think in that case we can compute `image` as the regular coimage,
--- i.e. the coequalizer of the kernel pair,
--- and that image has the appropriate mapping property.
-
--- def image_map {C C' : cochain_complex.{v} V} (f : C ⟶ C') (i : ℤ) :
---   image (C.d i) ⟶ image (C'.d i) :=
--- sorry
-
--- -- I'm not certain what the minimal assumptions required to prove the following
--- -- lemma are:
--- lemma induced_maps_commute {C C' : cochain_complex.{v} V} (f : C ⟶ C') (i : ℤ) :
--- image_to_kernel_map C i ≫ kernel_map f (i+1) =
---   image_map f i ≫ image_to_kernel_map C' i :=
--- sorry
+@[simp, reassoc]
+lemma image_to_kernel_map_condition (C : cochain_complex V) (i : ℤ) :
+  image_to_kernel_map C i ≫ kernel.ι (C.d (i + 1)) = image.ι (C.d i) :=
+by erw [limit.lift_π, fork.of_ι_app_zero]
+
+lemma induced_maps_commute [has_image_maps.{v} V] {C C' : cochain_complex.{v} V} (f : C ⟶ C')
+  (i : ℤ) :
+  image_to_kernel_map C i ≫ kernel_map f (i + 1) = image_map f i ≫ image_to_kernel_map C' i :=
+by { ext, simp }
 
 variables [has_cokernels.{v} V]
 
 /-- The `i`-th cohomology group of the cochain complex `C`. -/
 def cohomology (C : cochain_complex V) (i : ℤ) : V :=
 cokernel (image_to_kernel_map C (i-1))
 
--- TODO:
-
--- As noted above, as we don't get induced maps on boundaries with this generality,
--- we can't assemble the cohomology groups into a functor. Hopefully, however,
--- the commented out code below will work
--- (with whatever added assumptions are needed above.)
-
--- def cohomology_map {C C' : cochain_complex.{v} V} (f : C ⟶ C') (i : ℤ) :
---   C.cohomology i ⟶ C'.cohomology i :=
--- cokernel.desc _ (kernel_map f (i-1) ≫ cokernel.π _)
--- begin
---   rw [←category.assoc, induced_maps_commute, category.assoc, cokernel.condition],
---   erw [has_zero_morphisms.comp_zero],
--- end
-
--- /-- The cohomology functor from chain complexes to `ℤ` graded objects in `V`. -/
--- def cohomology_functor : cochain_complex.{v} V ⥤ graded_object ℤ V :=
--- { obj := λ C i, cohomology C i,
---   map := λ C C' f i, cohomology_map f i,
---   map_id' := sorry,
---   map_comp' := sorry, }
+variables [has_image_maps.{v} V]
+
+/-- A morphism of cochain complexes induces a morphism in cohomology at every degree. -/
+def cohomology_map {C C' : cochain_complex.{v} V} (f : C ⟶ C') (i : ℤ) :
+  C.cohomology i ⟶ C'.cohomology i :=
+cokernel.desc _ (kernel_map f (i - 1 + 1) ≫ cokernel.π _)
+begin
+  rw [←category.assoc, induced_maps_commute, category.assoc, cokernel.condition],
+  erw [has_zero_morphisms.comp_zero],
+end
```

Comment:
It looks like something may have improved with Lean 3.8.

1. The last `erw` can just be a `rw`.
2. If we add `@[reassoc]` to `induced_maps_commute`, then the whole `begin ... end` block can be `simp [induced_maps_commute_assoc]`.
3. If we add `@[simp, reassoc]` to `induced_maps_commute`, then the whole `begin ... end` block can just be `(by simp)`.

(I really like having the simplifier do everything, when it makes sense! It makes life so much easier for people who come later.)

If we do add `simp` to `induced_maps_commute`, we should think if there's any reason to prefer one direction of the simp lemma over the other, and even if we make an arbitrary choice, perhaps we can include a section the module doc explaining the "simp normal forms" we're aiming for. (I can do this if you agree it's a good change, if you prefer.)